sentence transformer blog (#333)
tembo-io/website: docusaurus/blog/2024-02-01-sentence-transformers/index.md

---
slug: sentence-transformers

> title: 'Automated Postgres Vector Search with pg_vectorize and HuggingFace sentence-transformers'

Reviewer: ```suggestion
title: 'Automated Postgres Vector Search with pg_vectorize and HuggingFace sentence transformers'
```

authors: [adam]
tags: [postgres, extensions, stacks, vector-search]
image: './tembo_vector.png'

> date: 2024-02-01T12:00

Reviewer: Adjust to publish time UTC when you're ready.

> description: Automated Vector Search workflow on Postgres using Hugging Face sentence-transformers and pg_vectorize

Reviewer: Should this be longer? Seems to just repeat the title

Author: Maybe we just remove it? Unless it helps with SEO?

Reviewer: Descriptions are often used for feed summaries and shorter lists of links (e.g., [justatheory.com/archive](https://justatheory.com/archive/) lists every post with the relevant description. Dunno if it's used for the Tembo blog, but if it is, or is in the future, could be useful.

---

![tembo-vector](./tembo_vector.png "tembo-vector")

> Sentence transformers have revolutionized text processing in machine learning and AI by converting raw text into numerical vectors, enabling applications like vector similarity searches and semantic search. With [pg_vectorize](https://github.com/tembo-io/pg_vectorize), an open source extension for Postgres, you have a seamless link to do vector search using over 120 open-source HuggingFace transformers, directly from Postgres.

Reviewer: ```suggestion
Sentence transformers have revolutionized text processing in machine learning by converting raw text into numerical vectors, enabling applications like vector similarity searches and semantic search. With [pg_vectorize](https://github.com/tembo-io/pg_vectorize), an open source extension for Postgres, you have a seamless link to do vector search using over 120 open-source HuggingFace transformers, directly from Postgres.
```

Reviewer: What is HuggingFace? Should there be a link?

Reviewer: I assume AI is in here and throughout for the keyword juice, but just to be pedantic, none of this is AI, just ML.</rant>

Author: Will provide a link to HF.

Can debate for years whether something is AI or just ML. There's definitely neural networks involved in this, which many would say is more AI than classical supervised/unsupervised machine learning.

pg_vectorize is completely open source, and you can run it locally or in your own environment by following the project’s [quick-start guide](https://github.com/tembo-io/pg_vectorize?tab=readme-ov-file#pg_vectorize). It is also available by default on the Tembo [VectorDB Stack](https://tembo.io/docs/tembo-stacks/vector-db) in [Tembo Cloud](https://tembo.io) and can be added to any Tembo Postgres instance.

## What are sentence-transformers?

> Machine learning and artificial intelligence models ultimately perform mathematical operations. So if you have raw text and want to build a machine learning or AI application, you need to find a way to transform text to numbers. There has been detailed [research](https://arxiv.org/abs/1908.10084) in this field. and  these transformer models are language models used for encoding the meaning behind the raw text data into an array of floats. Once you have the array of floats (also called vectors or embeddings), now you can do mathematical operations on that such as vector similarity search and other forms of machine learning. 

Reviewer: ```suggestion
Machine learning and artificial intelligence models ultimately perform mathematical operations. So if you have raw text and want to build a machine learning or AI application, you need to find a way to transform text to numbers. There has been detailed [research](https://arxiv.org/abs/1908.10084) in this field, and  the resulting transformer models encode the meaning behind the raw text data into an array of floats. Once you have the array of floats (also called vectors or embeddings), now you can perform efficient mathematical operations such as vector similarity search and other forms of machine learning. 
```

Author: good suggestion, thanks!

## Picking the right sentence transformer for your use case

> While OpenAI’s embedding API endpoint is commonly used, there are over [120 open-source](https://huggingface.co/sentence-transformers) sentence transformers available on HuggingFace. The right model to use is highly dependent on your data and your use case. For example, the [all-MiniLM-L12-v2](https://huggingface.co/sentence-transformers/all-MiniLM-L12-v2) model is a general purpose transformer, and can be used for a wide variety of use cases such as training a supervised classification model, or doing similarity search. Conversely, the [multi-qa-MiniLM-L6-dot-v1](https://huggingface.co/sentence-transformers/multi-qa-MiniLM-L6-dot-v1) model was built specifically for semantic search.

Reviewer: ```suggestion
While OpenAI’s embedding API endpoint is commonly used, there are over [120 open-source](https://huggingface.co/sentence-transformers) sentence transformers available on HuggingFace. The right model to use depends on your data and your use case. For example, the [all-MiniLM-L12-v2](https://huggingface.co/sentence-transformers/all-MiniLM-L12-v2) model is a general purpose transformer useful for a wide variety of use cases such as training a supervised classification model or performing similarity search. Conversely, the [multi-qa-MiniLM-L6-dot-v1](https://huggingface.co/sentence-transformers/multi-qa-MiniLM-L6-dot-v1) model was built specifically for semantic search.
```

> Model selection is a large area of research, so experimenting with transformers is highly recommended. Different models have different constraints on inputs, and produce different dimensionality of outputs. Theoretically, larger outputs can encode more information, but that is an area of active research. The tradeoff with larger dimensionality embeddings is that they require more storage and longer query times. You can  Read more about these pre-trained transformers [here](https://www.sbert.net/docs/pretrained_models.html).

Reviewer: Suggest using a more descriptive link text than "here", as in `Read more in [sbert's list of pretrained models](https://www.sbert.net/docs/pretrained_models.html).`

Author: good idea

## Direct SQL hooks to HuggingFace sentence transformers via pg_vectorize

> Pg_vectorize allows you to use any of Hugging Face’s sentence transformer models via SQL.t can be configured by simply changing the `model_name` parameter in the `transform_embeddings` function. To use this the [all-MiniLM-L12-v2](https://huggingface.co/sentence-transformers/all-MiniLM-L12-v2) model:

Reviewer: ```suggestion
pg_vectorize allows you to use any of Hugging Face’s sentence transformer models via SQL. It can be configured by simply changing the `model_name` parameter in the `transform_embeddings` function. For example, to use this the [all-MiniLM-L12-v2](https://huggingface.co/sentence-transformers/all-MiniLM-L12-v2) model:
```

```sql
SELECT * FROM vectorize.transform_embeddings(
   input => 'the quick brown fox jumped over the lazy dogs',
   model_name => 'sentence-transformers/all-MiniLM-L12-v2'
);
```

```text
{-0.0028378579299896955,0.055683549493551254,-0.025058165192604065,0.02842593938112259, ..., 0.0828385129570961}
```

> And to use the [all-mpnet-base-v2](https://huggingface.co/sentence-transformers/all-mpnet-base-v2) model, we just change the value of the `model_name` parameter.

Reviewer: ```suggestion
To use the [all-mpnet-base-v2](https://huggingface.co/sentence-transformers/all-mpnet-base-v2) model, just change the value of the `model_name` parameter.
```

```sql
SELECT * FROM vectorize.transform_embeddings(
   input => 'the quick brown fox jumped over the lazy dogs',
   model_name => 'sentence-transformers/all-mpnet-base-v2'
);
```

```text
{-0.028050078079104424,-0.004174928646534681,-0.001258305972442031,-0.017915889620780945, ..., 0.011784635484218596}
```

## Automate vector search workflow with just 2 SQL functions

Generating embeddings is one thing, but conducting vector search requires many steps:

> * Transform, store and index existing data in a table

Reviewer: ```suggestion
* Transform, store, and index existing data in a table
```

* Handle refreshes for any inserts or updates during the lifecycle of your application.

> * Transform raw text search queries into embeddings, using precisely the same transformer model,

Reviewer: ```suggestion
* Transform raw text search queries into embeddings, using precisely the same transformer model.
```

* And lastly, conduct a similarity search on the fly.

> This logic is typically left up to application developers and requires many tools to implement, but with pg_vectorize this becomes just two function calls; `vectorize.table()` and `vectorize.search()`

Reviewer: ```suggestion
This logic is typically left up to application developers and requires many tools to implement, but pg_vectorize enables it with just two function calls; `vectorize.table()` and `vectorize.search()`.
```

### Simplify vector creation and refresh via vectorize.table

> `vectorize.table()` is how you set up a vector search job. pg_vectorize keeps track of which transformer model you are using, and continuously monitors your table for changes. We select which model we want to use for embedding transformation by changing the value of the `transformer` parameter. pg_vectorize handles creates a column on your table to store embeddings for each row.

Reviewer: ```suggestion
`vectorize.table()` is how you set up a vector search job. pg_vectorize keeps track of the transformer model you use, and continuously monitors your table for changes. We select which model we want to use for embedding transformation by changing the value of the `transformer` parameter. pg_vectorize creates a column on your table to store embeddings for each row.
```

Reviewer: Decide whether to use `you` or `we` when describing actions to take.

```sql
SELECT vectorize.table(
   job_name => 'my_search_job',
   table => 'mytable',
   primary_key => 'record_id',
   transformer => 'paraphrase-MiniLM-L6-v2',
   schedule => 'realtime'
);
```

> By default, inserts and updates trigger an immediate transform of the new data using Postgres’s built-in [triggers](https://www.postgresql.org/docs/current/sql-createtrigger.html). However, pg_vectorize provides the flexibility to schedule updates using cron syntax by changing the `schedule` parameter. For example, `schedule -> ‘* * * * *’` will trigger a batch job that checks for updates or inserts every minute. All triggers result in a job enqueued to a [pgmq](https://github.com/tembo-io/pgmq) job queue, which means the transformation of text to embeddings does not slow down your insert or update events. Embedding transformations are always decoupled from the insert or update events.

Reviewer: ```suggestion
By default, inserts and updates trigger an immediate transform of the new data using Postgres’s built-in [triggers](https://www.postgresql.org/docs/current/sql-createtrigger.html). However, pg_vectorize provides the flexibility to schedule updates by changing the cron-syntax `schedule` parameter. For example, `schedule -> ‘* * * * *’` will trigger a batch job that checks for updates or inserts every minute. All triggers send a job to a [pgmq](https://github.com/tembo-io/pgmq) job queue, which means the transformation of text to embeddings does not slow down insert or update events. Embedding transformations are always decoupled from the insert or update events.
```

![vectorize-table](./vectorize-table.png "vectorize-table")

### Intuitive vector search using vectorize.search

> `vectorize.search()` is how you query your data given a raw text query. To search `mytable` for something like “mobile electronic accessories'', we need to transform our raw text query into embeddings, being certain to use precisely the same sentence-transformer that we used to generate the embeddings that are written to our database. Then, using those embeddings, conduct a vector similarity search via [pgvector](https://github.com/pgvector/pgvector) and return the results of the query. This orchestration is often handled within an application built by the dev team, but since this is just a feature in the database triggered by a function call, we no longer have to build it.

Reviewer: ```suggestion
`vectorize.search()` is how you query your data given a raw text query. To search `mytable` for something like “mobile electronic accessories”, we need to transform the raw query into embeddings, being certain to use precisely the same sentence-transformer that generated the embeddings written to our database. Then, using those embeddings, conduct a vector similarity search via [pgvector](https://github.com/pgvector/pgvector) and return the results of the query. This orchestration is often handled within an application built by the dev team, but since this is just a feature in the database triggered by a function call, we no longer have to build it.
```

```sql
SELECT * FROM vectorize.search(
   job_name => 'my_search_job',
   query => 'mobile electronics accessories',
   return_columns => ARRAY['product_id', 'product_name'],
   num_results => 5
);
```

Results are returned as rows of jsonb, including a similarity score for the search query, sorted in descending order.

```text
                                        search_results                                        
------------------------------------------------------------------------------------------------
{"product_id": 13, "product_name": "Phone Charger", "similarity_score": 0.791816648695175}
{"product_id": 23, "product_name": "Flash Drive", "similarity_score": 0.7456739345117277}
{"product_id": 4, "product_name": "Bluetooth Speaker", "similarity_score": 0.7386882311199485}
```

## Separate compute for transformer models

> Machine learning workloads typically have very different compute requirements than Postgres. For example, many transformers will benefit immensely from having a GPU available, or massive amounts of CPU and memory. Therefore, it makes sense to run the embedding transformers on a separate host than Postgres itself, which does not natively require a GPU to be present. So in our example, when we run “[docker compose up](https://github.com/tembo-io/pg_vectorize/blob/main/docker-compose.yml)”, we’re running the transformer models in a separate container than Postgres. This gives you the flexibility to run the transformers wherever you want, so long as postgres can establish an HTTP connection with the host of the container running the transformers.

Reviewer: ```suggestion
Machine learning workloads typically have very different compute requirements than Postgres. For example, many transformers benefit immensely from a GPU, or massive amounts of CPU and memory. Therefore, it makes sense to run the embedding transformers on a separate host from Postgres itself, which does not natively require a GPU. So in our example, when we run [`docker compose up`](https://github.com/tembo-io/pg_vectorize/blob/main/docker-compose.yml), we’re running the transformer models in a separate container from Postgres. This gives you the flexibility to run the transformers wherever you want, so long as Postgres can establish an HTTP connection with the host of the container running the transformers.
```

Reviewer: HTTP connection? Or network?

Author: Yes. It is an HTTP connection over the docker network...

Author: It's HTTP requests from postgres to the container hosting the transformer models, is what I mean.

Reviewer: It just reads a little odd to me because Postgres itself doesn't use HTTP; I presume pg_vectorize does. In any event, maybe use "can connect to" instead.

Author: Ah yes, its pg_vectorize that opens the HTTP connection. Will clarify.

![containers](./containers.png "containers")

> In Tembo Cloud, we run this workload on Kubernetes and it is managed with the [tembo-operator](https://github.com/tembo-io/tembo/tree/main/tembo-operator) instead of `docker-compose`. This also gives us the flexibility to assign and change the compute that is allocated to the transformer container.

Reviewer: ```suggestion
In Tembo Cloud, we run this workload on Kubernetes and manage it with the [tembo-operator](https://github.com/tembo-io/tembo/tree/main/tembo-operator) instead of `docker-compose`. This gives us the flexibility to assign and change the compute allocated to the transformer container.
```

> The container hosting the transformers is simple; it is a python http server built with FastAPI that runs the [SentenceTransformer](https://pypi.org/project/sentence-transformers/) library. The general-purpose all-MiniLM-L12-v2 model is saved in the container image, so when the container starts, that model is immediately loaded into memory. When you call the container for any other model, it is downloaded and cached on the fly.

Reviewer: ```suggestion
The container hosting the transformers is a simple python http server built with FastAPI running the [SentenceTransformer](https://pypi.org/project/sentence-transformers/) library. It saves general-purpose all-MiniLM-L12-v2 model in the image, so when the container starts the model is immediately loaded into memory. When you call the container for any other model, it is downloaded and cached on the fly.
```
